### ISSUE TITLE
Support PostgreSQL Fulltext Search

### ISSUE THREAD

**Issue by jgillich** [mastodon/mastodon#26645]
### Pitch

Full-text search will become much more useful in the upcoming 4.2 release, but it requires ElasticSearch to function. ElasticSearch is known to be especially heavyweight, which is a challenge for many small to medium sized instances. As of right now, very few instances deploy it.

I propose adding Postgres-based search functionality as an alternative to ElasticSearch. It appears that native full-text search functionality in Postgres was not chosen due to [concerns with performance and disk usage](https://mastodon.social/@Gargron/99463924907814913). I've done a bit of research on this topic and I believe there could be a solution in the form of the [Trigram extension](https://www.postgresql.org/docs/current/pgtrgm.html) (included in the standard PG distribution). For an introduction, I'd recommend the blog post about how [Gitlab improved their search performance by switching from full-text search to Trigram indexes](https://about.gitlab.com/blog/2016/03/18/fast-search-using-postgresql-trigram-indexes/).

Trigram search is based on word similarity and do not require exact matches, which is preferable for most queries due to differences in spelling, punctuation, typos etc. And it is quite fast, for a deep dive into optimization read [Optimizing Postgres Text Search with Trigrams](https://alexklibisz.com/2022/02/18/optimizing-postgres-trigram-search). They use a dataset of 8.9M Amazon book reviews. The size of their index for the summary (title) column is 1GB, which seems reasonable, but they omitted the review body from this experiment. Some testing would need to be performed, but it's likely that the indexes would still be smaller than the space consumed by ElasticSearch.

Therefore, I think implementing search based on Trigram indexes could be a worthwhile addition. It would make full-text search available to everyone without the additional resource usage and setup/maintenance burden of ElasticSearch. If possible, it would be enabled via an environment variable, and indexes would not be created unless the feature is turned on. Not the this is not indended to replace ElasticSearch entirely since the quality of results is likely better with a proper search engine.

### Motivation

ElasticSearch adds a lot of cost and complexity to a setup, which matter greatly to operators of small and medium-sized instances. For example, masto.host charges between $5 and $10 for ElasticSearch, which can be a significant burden to the operators. But it's likely that search will become much more desirable in the future, therefore Mastodon should provide an option that is cheaper and easier to deploy.

**Comment by ch0ccyra1n** [comment]
Yes please! I run a smaller instance but would love to have fulltext search without spending extra money on it. It's hard to justify the expense on such a small instance (about 50 users), although as mentioned above, the feature is desirable and is likely to become more desirable in the future.

**Comment by nanos** [comment]
It's maybe worth noting that @VyrCossont has tried a PostgreSQL backed full text search before - I have no idea how that went though

https://github.com/VyrCossont/mastodon/pull/3

**Comment by VyrCossont** [comment]
That patch is really just a proof of concept, though it worked well enough on a small instance compared to the alternative of "no search at all".

A real implementation would include other text fields (polls and media) and strip HTML before indexing the text, likely into a dedicated special-purpose column or even table, plus it would tie into the search parser/transformer classes to generate a query from advanced search syntax.

**Comment by amdavidson** [comment]
I would bounty this one. This is a big win for small instances. Mastodon is already a bit heavyweight for a single user but ES is a beast.

**Comment by jgillich** [comment]
@VyrCossont Duplicating all status text to strip HTML would greatly increase the database size, I'm not sure if that's necessary. With a trigram `ILIKE` search query, you would drop words with fewer than 3 characters ([see Gitlab implementation](https://gitlab.com/gitlab-org/gitlab-foss/-/blob/5ed2199ea73700ec5a489b5d7d1c746c6b4f9d34/lib/gitlab/sql/pattern.rb#L42)), which eliminates most supported HTML elements

There are a couple of exceptions like `span`, `code` and  `del` or attributes like `href`, but the impact is minimal. Mastodon doesn't support authoring HTML posts, and Markdown in glitch-soc is federated as text/markdown, so that's probably what Mastodon will do as well. Not sure which servers support HTML, but if considering that Mastodon has about 80% market share, it seems plausible that fewer than 1% of statuses use HTML.

Searching for common HTML terms like `span` would likely include a number of irrelevant results, but I think that is a worthwhile tradeoff to keep storage costs down. Those who care about getting the best results will want to deploy ElasticSearch.

**Comment by Gargron** [comment]
The format of federated posts is HTML. All remote posts in your database are HTML.

**Comment by r4v5** [comment]
> Duplicating all status text to strip HTML would greatly increase the database size, I'm not sure if that's necessary.

Keep in mind that the current alternative implementation of fulltext search (ES) also requires the same duplication of ingested data, it's just in ES shards instead of PG tables.

**Comment by DanielHeath** [comment]
@jgillich Storage isn't the expensive part of running ES - CPU, RAM, and operator time are the expensive bit. Elasticsearch is designed with a default assumption of high scale, which means that it's very expensive to run at low scale.

The advantage of putting search into postgres is that it's operationally simple for small to mid sized instances - the core postgres distribution already includes the text search features, and the server is already running postgres.

I say this as someone running a search page off postgres - I'm getting about 1 search per second, and queries take less than 0.01 seconds each (my monitoring doesn't go lower than that).

**Comment by pointlessone** [comment]
I run a single-user instance and would like to provide some numbers for context.

My instance db:

|Table|Records|Size|
|-|-|-|
|statuses |662,135|341 MB|
|accounts|62,097|107 MB|
|tags|56,905|5.2 MB|

Total db size is 990 MB.
PG uses about 335 MB of RAM (resident).

Elasticsearch:

Index size on disk is 112 MB

I can not get it to use less than ~650 MB of RAM (resident) and ~3500 MB virtual.

My VPS has 4 GB of RAM and 80 GB of disk space. It's absolutely impossible to make Mastodon (web, sidekiq, and streaming), dbs (pg and redis), and Elasticsearch work without swap. Without Elasticsearch it can fit all in RAM but Elasticseach just hogs so much resources. With 4 GB of swap it barely fits. I would happily trade trice the db size and twice pg ram to get rid of Elasticsearch.

Also notice that my instance has over half a million statuses. This is more representative of a medium size instance rather than a single-user one. I do some non-standard things to my instance.

Every resource on the internet about Elasticsearch states that minimum requirement in hardware is 8GB of RAM with half of it going to Elasticsearch itself and the other half reserved for disk cache. This is utterly ridiculous scale for 112 MB index. Even 650 MB is baffling. My whole index can fit 5 times over in that space. I have no idea what it does with 4 GB of virtual memory.

**Comment by pointlessone** [comment]
As an alternative to consider, here's a Postgres extension that provides "Elastic-Quality Full Text Search": [pg_bm25](https://docs.paradedb.com/blog/introducing_bm25). It is new and how good it is is unclear yet but it looks promising compared to pg full-text search and it doesn't look like current implementation uses any advanced Elasticsearch features either. I suspect it might be much less resource-intensive and probably cheaper to deploy for both small and big instances.

**Comment by DanielHeath** [comment]
`pg_bm25` looks quite promising; although it's not supported on major hosts like RDS, small instances are likely running their own postgres anyways.
